bip-0331.mediawiki

```diff
@@ -0,0 +1,394 @@
+learning that it is the ancestor of the high-feerate transaction.</ref>
+
+[[File:./bip-0331/package_cpfp_flow.png|800px]]
+
+This protocol is receiver-initiated only; nodes do not proactively announce packages to their peers.
+<ref>'''Why no sender-initiated protocol?''' Sender-initiated package
+relay can, theoretically, save a round trip by notifying the receiver ahead of time that they will
+probably need to request and validate a group of transactions together in order for them to be
+accepted. As with any proactive communication, there is a chance that the receiver already knows
+this information, so this network bandwidth may be wasted. Shortened latency is less significant
+than wasted bandwidth.
+
+The logic used to decide when to announce a package proactively determines whether it is a net
+increase or decrease for overall bandwidth usage. However, it's difficult to design anything to save
+bandwidth without any idea of what its bandwidth usage actually looks like in practice. We'll want
+to design the sender-initiated protocol carefully, and inform the design decisions using data
+collected from the mainnet p2p network. However, there is no historical transaction data to use
+because the goal is to enable currently-rejected transactions to propagate. In order to get this
+right, I propose we hold off on sender-initiated for now, deploy receiver-initiated package relay,
+observe its usage and figure out where we can save a round trip, and then introduce a
+well-researched sender-initiated package relay protocol.</ref>
+
+Ancestor package relay is negotiated between two peers during the version handshake using a
+"sendpackages" message containing version=1. It requires both peers to support wtxid-based relay
+because package transactions are referenced by their wtxid. Peers wishing to relay multiple versions
+of packages should send one "sendpackages" message for each version.
+<ref>'''Is it possible to support other types of packages in the future?''' Yes. Attempting to
+support arbitrary packages in mempool validation may result in very complex logic, new Denial of
+Service attack vectors, and policy limitations that could be leveraged to censor transactions (aka
+"pinning attacks").  This protocol is extensible to support multiple types of packages based on
+future desired use cases.</ref>
+<ref>'''Why send one message per version instead of sending a single message for the highest
+supported version?''' It should also be possible to support some subset of the existing package
+types. For example, if a node's mempool policy doesn't support or a node implementation deprecates
+version 2, it should be easy to announce support for version 1, but not version 2, packages.</ref>
+
+[[File:./bip-0331/version_negotiation.png|600px]]
+
+===Combined Hash===
+
+A "combined hash" serves as a unique "package id" for some list of unique transactions.
```

Comment:
I think one use case to unambiguously determine if we've already validated (/rejected) a certain (exact) package as per [this comment](https://gist.github.com/glozow/8469dc9c3a003c7046033a92dd504329#:~:text=For%20packages%20and%20sub%2Dpackages%20that%20fail%2C%20add%20the%20transaction%20group%20by%20hashing%20the%20wtxids%20of%20each%20of%20the%20transactions%2C%20sorted%20lexicographically.%20Ensure%20that%20each%20group%20is%20itself%20an%20ancestor%20package.%20When%20a%20package%20is%20partially%20submitted%2C%20exclude%20the%20transactions%20that%20ended%20up%20in%20the%20mempool.)

Reply:
Yes, the combined hash is only used in `notfound(MSG_PKGTXNS)`.

We cannot use the child wtxid because the child wtxid is not included in the `getpkgtxns` request.